### ISSUE TITLE
Unable to install the package

### ISSUE THREAD

**Issue by sofianeh** [jupyterhub/jupyterlab-hub#2]
When I execute `pip install jupyterhub_labextension`, I receive the following error:
```
Could not find a version that satisfies the requirement jupyterhub_labextension (from versions: )
No matching distribution found for jupyterhub_labextension
```

**Comment by Analect** [comment]
@danielballan 
Thanks for your efforts with this.
Is this extension in a usable state against jupyterhub 0.72 and jupyterlab 0.12.1?
Does it need to be installed on both jupyterhub and jupyterlab or just the hub?

**Comment by danielballan** [comment]
@sofianeh This package is not pip-installable yet because it is still a work in progress, not quite functional. I expect it will be released as soon as it is minimally functional.

@Analect It's not quite in a usable state yet (see comments in my PR).

**Comment by Analect** [comment]
Thanks @danielballan 
I see [this](https://github.com/jupyterlab/jupyterlab/issues/1375) was fixed around the time of your last pull request, which I think is related.  Is the plan for the jupyterlab extension system to bed-down some more before revisiting this extension?

**Comment by danielballan** [comment]
@Analect I'd like to get this working with the latest stable notebook server + current JupyterLab to start playing with it. Maybe we'll wait to tag a release until we have an approach that we plan to sustain.

**Comment by Analect** [comment]
@danielballan That would be great.  I have a set-up with 0.14 jupyterlab and ver 5.0-dev (master branch) of notebook.  If you have some guidance as to how to install this extension and whether it needs installing on both jupyterhub and jupyterlab, that would be really helpful.

**Comment by Analect** [comment]
@danielballan ... any guidance on whether this can work in the context of jupyterlab 0.16.2? I don't know enough about how these lab extensions work, but was this [pull request](https://github.com/jupyterlab/jupyterlab/pull/1541) a step towards getting this functional?

**Comment by danielballan** [comment]
I think that PR is unrelated. But I'm (finally) revisiting this right now, armed with better debugging tools. Will report back shortly.

**Comment by danielballan** [comment]
@Analect Ah, you were right that it is related. Going back a littler further in time, [this commit](https://github.com/jupyterlab/jupyterlab/pull/1071/commits/df30cfd26cbdf0c49f092372ab2869717cf35f0e) cleaning up the config, superceding [this commit](https://github.com/jupyterlab/jupyterlab/pull/1071/commits/df30cfd26cbdf0c49f092372ab2869717cf35f0e) that first made it possible for hub to inject its URLs. I'm refactoring the extension to fit....

**Comment by danielballan** [comment]
Update: https://github.com/jupyterlab/jupyterlab/pull/1804

**Comment by Analect** [comment]
@danielballan 
Thanks for your work on this. With that pull request above (1804) already merged, is jupyterlab now set-up to be able to work with this extension, or do some things still need modifying.  I'd love to see a worked example of what's possible.

**Comment by danielballan** [comment]
@Analect See #4.

**Comment by Analect** [comment]
@danielballan 
Thanks .. I hadn't seen that pull request. I look forward to testing when I see jupyterlab 0.17 arriving.
Could I ask a few follow-on questions.

The image I currently use to spin up a jupyterlab server uses this command:

```
CMD jupyterhub-singleuser \
  --port=8888 \
  --ip=0.0.0.0 \
  --user="$JPY_USER" \
  --cookie-name=$JPY_COOKIE_NAME \
  --base-url=$JPY_BASE_URL \
  --hub-prefix=$JPY_HUB_PREFIX \
  --hub-api-url=$JPY_HUB_API_URL \
  --HubAuth.api_token=$JPY_API_TOKEN \
  --notebook-dir="/home/jovyan/work" \
  --SingleUserNotebookApp.default_url=/lab
```
... and I am using kubespawner to spawn servers, so my `jupyterhub_config.py` includes the following elements ... as an illustration of how parameters are passed into that spawner.

```
from kubespawner.spawner import KubeSpawner
class MySpawner(KubeSpawner):
... 

c.JupyterHub.spawner_class = MySpawner
c.MySpawner.namespace = os.environ.get('POD_NAMESPACE', 'default')
```

I can see from the docs in your pull request that you suggest adding `c.Spawner.cmd = ['jupyterhub-singlelabuser']` into jupyterhub_config.py.  I'm just wondering how that will work in conjunction with the kubespawner I am using ... can a `c.Spawner.*` parameter work in conjunction with a custom `c.MySpawner.*` (my spawner, which is leveraging the kubespawner, as above).

Or perhaps it is enough to install this extension on any image that is being used to spawn a server .. and the command will contain something like:

```
CMD jupyterhub-singlelabuser \
--parameterX
--parameterY
...
```
Thanks for your insights.

**Comment by danielballan** [comment]
Yes, I think `CMD jupyterhub-singlelabuser \` should do it. I will also be running a custom spawner and I _expect_ this to interact with that just fine... if not I'll report back with more instructions in the README.

If you are eager to try this out, you can install the 0.17.0-0 pre-release.

**Comment by Analect** [comment]
@danielballan 
I was trying to install the development version of the labextension (since I don't think `pip install jupyterhub_labextension` is wired up yet).

Below is an extract from part of a Dockerfile that I use to do these various installs, preceeded by installs of jupyterhub and jupyterlab (although I'm not sure if your labextension install .. which also has jupyterlab as a dependency ... negates the need for the explicit jupyterlab install first)?

I forked your repo, just to correct the reference to `jupyterlab 0.17.0-0` (extra `-0`) which was preventing the `npm install` from running, hence the `git clone -b revisited https://github.com/Analect/jupyterhub-labextension` below.

```
WORKDIR /home/$NB_USER
RUN pip install --no-cache-dir --pre jupyterhub
RUN conda install -c conda-forge jupyterlab --yes
RUN jupyter serverextension enable --py jupyterlab --sys-prefix
# add labextension
RUN git clone -b revisited https://github.com/Analect/jupyterhub-labextension
WORKDIR /home/$NB_USER/jupyterhub-labextension
RUN npm install
RUN pip install -e .
RUN jupyter labextension install --symlink --py --sys-prefix jupyterhub_labextension
RUN jupyter labextension enable --py --sys-prefix jupyterhub_labextension
```

I hit these glitches though (link below).  Is it best just to wait for the next jupyterlab release, or can you suggest an easy fix?
https://gist.github.com/Analect/0ea300a9bab5f4ca45c50d70f0bf9041

Thanks.

**Comment by danielballan** [comment]
It looks like you are getting the 0.16.2 verison of the nodejs packages. Given the complexity of debugging through Docker, I would try to get this running directly first. Or wait for 0.17.0 -- any day now.

**Comment by Analect** [comment]
@danielballan 
Ran this again this morning against jupyterlab 0.17.0, which has been released.  Still showing problems with `Cannot find module '@jupyterlab/services/lib'`.  Any thoughts on how I might resolve this?

```
Step 26/38 : RUN jupyter serverextension enable --py jupyterlab --sys-prefix
 ---> Running in ab2e7a5c87f3
Enabling: jupyterlab
- Writing config: /opt/conda/etc/jupyter
    - Validating...
      jupyterlab  OK
 ---> 6a5aa3795083
Removing intermediate container ab2e7a5c87f3
Step 27/38 : RUN git clone -b revisited https://github.com/danielballan/jupyterhub-labextension
 ---> Running in 7882fb0f7f00
Cloning into 'jupyterhub-labextension'...
 ---> bddf0390f1dd
Removing intermediate container 7882fb0f7f00
Step 28/38 : WORKDIR /home/$NB_USER/jupyterhub-labextension
 ---> b64cc4347b1d
Removing intermediate container a1c1dec9469d
Step 29/38 : RUN npm install
 ---> Running in 40731a4c78fc
npm WARN optional dep failed, continuing fsevents@1.1.1

> jupyterlab@0.17.0 postinstall /home/jovyan/jupyterhub-labextension/node_modules/jupyterlab
> node scripts/dedupe.js

npm WARN unavoidable conflict domelementtype { '/home/jovyan/jupyterhub-labextension/node_modules/jupyterlab/node_modules/sanitize-html/node_modules/htmlparser2/node_modules/domutils/node_modules/dom-serializer/node_modules/domelementtype': '1.1.3',
npm WARN unavoidable conflict   '/home/jovyan/jupyterhub-labextension/node_modules/jupyterlab/node_modules/sanitize-html/node_modules/htmlparser2/node_modules/domelementtype': '1.3.0' }
npm WARN unavoidable conflict Not de-duplicating

> jupyterhub_labextension@0.1.0 prepublish /home/jovyan/jupyterhub-labextension
> npm run build


> jupyterhub_labextension@0.1.0 build /home/jovyan/jupyterhub-labextension
> npm run build:src && npm run build:extension


> jupyterhub_labextension@0.1.0 build:src /home/jovyan/jupyterhub-labextension
> tsc --project src && node scripts/copyfiles.js

src/plugin.ts(10,8): error TS2307: Cannot find module '@jupyterlab/services/lib'.
src/plugin.ts(71,23): error TS2345: Argument of type '{ commands: CommandRegistry; }' is not assignable to parameter of type 'IOptions'.
  Types of property 'commands' are incompatible.
    Type 'CommandRegistry' is not assignable to type 'CommandRegistry'. Two different types with this name exist, but they are unrelated.
      Types of property 'commandChanged' are incompatible.
        Type 'ISignal<CommandRegistry, ICommandChangedArgs>' is not assignable to type 'ISignal<CommandRegistry, ICommandChangedArgs>'. Two different types with this name exist, but they are unrelated.
          Types of property 'connect' are incompatible.
            Type '(slot: Slot<CommandRegistry, ICommandChangedArgs>, thisArg?: any) => boolean' is not assignable to type '(slot: Slot<CommandRegistry, ICommandChangedArgs>, thisArg?: any) => boolean'. Two different types with this name exist, but they are unrelated.
              Types of parameters 'slot' and 'slot' are incompatible.
                Type 'Slot<CommandRegistry, ICommandChangedArgs>' is not assignable to type 'Slot<CommandRegistry, ICommandChangedArgs>'. Two different types with this name exist, but they are unrelated.
                  Types of parameters 'sender' and 'sender' are incompatible.
                    Type 'CommandRegistry' is not assignable to type 'CommandRegistry'. Two different types with this name exist, but they are unrelated.
```

**Comment by danielballan** [comment]
I just made a fresh environment and confirmed that this works for me:

```bash
$ conda create -n test-jhl
$ source activate test-jhl
$ cd Documents/Repos/jupyterhub-labextension
$ pip install .
$ conda list | grep jupyter
jupyter-client            5.0.0                     <pip>
jupyter-core              4.3.0                     <pip>
jupyterhub                0.7.2                     <pip>
jupyterhub-labextension   0.1.0                     <pip>
jupyterlab                0.17.0                    <pip>
$ cd
$ jupyter labextension install --symlink --py --sys-prefix jupyterhub_labextension
$ jupyter labextension enable --py --sys-prefix jupyterhub_labextension
$ jupyter labextension list
Known labextensions:
config dir: /Users/dallan/miniconda/envs/test-jhl/etc/jupyter/labconfig
    jupyterhub_labextension enabled : /Users/dallan/miniconda/envs/test-jhl/share/jupyter/labextensions/jupyterhub_labextension
jupyterhub --config=jupyterhub_config.py
```

The error you are seeing comes from having conflicting versions of the JS. I recommend trying to debug this outside of Docker.

**Comment by Analect** [comment]
@danielballan 
I managed to get this installed via Docker ... I had been using a `nodejs` version installed from `conda` and that was what was causing the problem.

As per further up this thread [here](https://github.com/jupyterhub/jupyterhub-labextension/issues/2#issuecomment-282763814), I've installed jupyterhub, jupyterlab and this lab extension (development install) into a docker image that is then used to spawn a notebook server ...  that's the approach that has worked fine for me up to now ... previously with the `CMD jupyterhub-singleuser` followed by several parameters ... that image was able to spawned by the hub.

I was hoping to have just replaced `CMD jupyterhub-singleuser` with `CMD jupyterhub-singlelabuser`, but with those various installs above, this doesn't work.  I then tried with `CMD python /home/jovyan/jupyterhub-labextension/cli_scripts/jupyter-singlelabuser` ... hoping that by being more explicit where that command tool was would work, but it doesn't.

Am I missing something else?  Does this lab extension also need to be installed on the hub itself?

**Comment by danielballan** [comment]
As far as I can tell you're not missing anything. Refer back to my last comment showing the complete list of steps. Nothing has to be done with the hub other than ensure that `jupyter-singlelabuser` replaces `jupyter-singleuser`. More detail about exactly how it's failing would help.

**Comment by danielballan** [comment]
Yep, it's the same thing. You have the right idea. It should get added to your PATH [by this line in the setup.py](https://github.com/jupyterhub/jupyterhub-labextension/blob/master/setup.py#L26) when the Python package jupyterhub_labextension is installed.

**Comment by Analect** [comment]
OK ... that somehow isn't working.  What is the best way for me to do this 'manually' or more explicitly?

**Comment by danielballan** [comment]
I would start by trying to reproduce the steps above exactly, sans Docker, just to be sure they work as expected on your system.

**Comment by Analect** [comment]
I think I see the problem ... the command `jupyter-singlelabuser` works ... but not `jupyterhub-singlelabuser` as per your README.  ie. .. you are specifying that extra **hub** part in the command, which is incorrect ... is that right?

**Comment by danielballan** [comment]
Bah. You're correct. Sorry about that. I'll add a PR to fix now....

**Comment by Analect** [comment]
No problem.  Glad we got to the bottom of it. Let me try to spin things up now.

**Comment by Analect** [comment]
@danielballan 
Got things working, with the 'Hub' menu item now appearing (1 below).
I was wondering if there was a way to leverage this approach with the lab extension to be able to add various tool buttons in a band along the top of the jupyterlab (red rectangle) ... as is the case with the classic tree view (2 below)?

![image](https://cloud.githubusercontent.com/assets/4063815/23514709/09b3f6f4-ff60-11e6-87d0-0411e381f6ca.png)

Below is something that the azure team have been able to do (also in the context of the classic view), but I'm trying to establish if the same thing is possible in jupyterlab now, with this extension?

![image](https://cloud.githubusercontent.com/assets/4063815/23514940/bc650fcc-ff60-11e6-9ae3-72ee5dc0042b.png)

**Comment by danielballan** [comment]
I don't think that the plugin architecture supports that. I imagine that multiple plugins might interact badly if they are all trying to add HTML to the top area. The menus, by contrast, are nicely composable. However, you can do custom branding by customizing `lab.html` directly. This is just my guess -- you can post on the JupyterLab issues directly to confirm that with a core dev.

**Comment by Analect** [comment]
Great. Thanks for clarifying.

**Comment by Analect** [comment]
@danielballan 
Have you been keeping an eye on the new extension mechanism in jupyterlab.  I'm finding it hard to follow things, but was wondering if your approach still works with jupyterlab 0.20.0 and above?  I noticed there was some discussion around updating it on the jupyterlab gitter channel around [here](https://gitter.im/jupyterlab/jupyterlab?at=5900d2bfd32c6f2f09429004) ... but I'm not sure that ever materialised.

**Comment by danielballan** [comment]
0.21.0 was tagged yesterday. The core devs are in the process of pushing up all the related packages. I will open a PR shortly targeting 0.21.0.

**Comment by Analect** [comment]
Thanks @danielballan  ... will look out for it.

**Comment by ellisonbg** [comment]
Have updated the repo to work with 0.21. However, the python CLI stuff has been moved to JupyterLab (see the README of this repo for details). We should be releasing JupyterLab 0.22 soon, and then we will need to update a few APIs in this repo. I will open a new issue for that. Closing this one for now...
